examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/model/repositories/ApiRepository.java

```diff
@@ -0,0 +1,9 @@
+package uk.gov.api.springboot.domain.model.repositories;
+
+import java.util.List;
+import uk.gov.api.springboot.domain.model.Api;
+
+public interface ApiRepository {
```

Comment:
:question: this may not make sense in this Ring https://herbertograca.com/2017/09/21/onion-architecture/#comment-25505

Reply:
The thought I was having was whether we'd want:

```java
package uk.gov.api.springboot.domain.model;

import java.util.List;

public interface Storage {
    List<Api> findAll();
}
```

And it'd be implemented in `Application`.

**Or** would we want `Storage` to be implemented in the domain, in which case we'd want to have an `interface ApiRepository` here, so we can implement it in the `Application` layer?

Reply:
What tradeoffs do you see with each route?

Reply:
Firstly, not sure what the "recommended" way of doing it is if doing this architecture / trying to implement Domain Driven Design.

Given the idea of having the Domain layer is that it's core business rules that can be nicely tested independently, I'd lean towards the `Storage` being implemented in the Domain layer, which relies on an `interface ApiRepository` in the Domain Models, which means:

- `Storage`'s key functionality is built + tested in the Domain ring
- `ApiRepository` can be injected in, via the constructor
- `ApiRepository` can be implemented in the Infrastructure ring, and in a specific way that makes sense for the app, and could use i.e. a specific DB implementation

Alternatively, if `Storage` was not implemented in the Domain layer, we'd:

- have something core to the Domain not being implemented and tested in the Domain, potentially losing control / visibility over business rules, requiring `Storage` would need a Domain Service to be injected to manage any business rules

I think it makes sense to have the Domain Models contain a `Storage`, as a concrete implementation, which depends on an `interface ApiRepository`, which can then be implemented outside of the Domain.

Reply:
I think you answered the question yourself?

Reply:
I'm trying to get my head round having the storage implementation in the domain layer - I thought it would be an implementation of the persistence layer, so wouldn't it go in the infrastructure ring? Or am I misunderstanding what we mean by storage here?

Reply:
So this isn't the implementation itself, only an `interface`. This allows us to inject in the `interface` to the implementation of `Storage`, while keeping `ApiRepository`'s implementation clear to implement as you mention in the Infrastructure ring

Reply:
The bit I still don't understand is "the implementation of Storage (Domain)" - that's the bit I thought would go in infrastructure?

Reply:
So `Storage` itself can be implemented in the Domain - we can set up the business rules for what needs to be validated before persisting/how to retrieve entries, but the underlying _how that thing is actually stored_ can be injected in

Reply:
that does make sense, thanks